test/types/community/collection/findX.test-d.ts

```diff
@@ -300,3 +300,223 @@ expectAssignable<SchemaWithUserDefinedId | null>(await schemaWithUserDefinedId.f
+interface Foo {
+  a: string;
+}
+
+const fooObj: Foo = {
+  a: 'john doe'
+};
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const fooFilter: Filter<Foo> = fooObj;
+
+// Specifically test that arrays can be included as a part of an object
+//  ensuring that a bug reported in https://jira.mongodb.org/browse/NODE-3856 is addressed
+interface FooWithArray {
+  a: number[];
+}
+
+const fooObjWithArray: FooWithArray = {
+  a: [1, 2, 3, 4]
+};
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+const fooFilterWithArray: Filter<FooWithArray> = fooObjWithArray;
+
+declare const coll: Collection<{ a: number; b: string }>;
+expectType<WithId<{ a: number; b: string }> | null>((await coll.findOneAndDelete({ a: 3 })).value);
+expectType<WithId<{ a: number; b: string }> | null>(
+  (await coll.findOneAndReplace({ a: 3 }, { a: 5, b: 'new string' })).value
+);
+expectType<WithId<{ a: number; b: string }> | null>(
+  (
+    await coll.findOneAndUpdate(
+      { a: 3 },
+      {
+        $set: {
+          a: 5
+        }
+      }
+    )
+  ).value
+);
+
+// projections do not change the return type - our typing doesn't support this
+expectType<WithId<{ a: number; b: string }> | null>(
+  (await coll.findOneAndDelete({ a: 3 }, { projection: { _id: 0 } })).value
+);
+
+/**
+ * mutually recursive types are not supported and will not get type safety
+ */
+interface A {
+  b: B;
+}
+
+interface B {
+  a: A;
+}
+
+declare const mutuallyRecursive: Collection<A>;
+//@ts-expect-error
+mutuallyRecursive.find({});
+mutuallyRecursive.find({
+  b: {}
+});
+
+/**
+ * types that are not recursive in name but are recursive in structure are
+ *   still supported
+ */
+interface RecursiveButNotReally {
+  a: { a: number; b: string };
+  b: string;
+}
+
+declare const recursiveButNotReallyColl: Collection<RecursiveButNotReally>;
+expectError(
+  recursiveButNotReallyColl.find({
+    'a.a': 'asdf'
+  })
+);
+recursiveButNotReallyColl.find({
+  'a.a': 2
+});
+
+/**
+ * recursive schemas are now supported, but with limited type checking support
+ */
+interface RecursiveSchema {
+  name: RecursiveSchema;
+  age: number;
+}
+
+declare const collection3: Collection<RecursiveSchema>;
+collection3.find({
+  name: {
+    name: {
+      age: 23
+    }
+  }
+});
+
+collection3.find({
+  age: 23
+});
+
+/**
+ * Recursive optional schemas are also supported with the same capabilities as
+ *   standard recursive schemas
+ */
+interface RecursiveOptionalSchema {
+  name?: RecursiveOptionalSchema;
+  age: number;
+}
+
+declare const collection4: Collection<RecursiveOptionalSchema>;
+
+collection4.find({
+  name: {
+    name: {
+      age: 23
+    }
+  }
+});
+
+collection4.find({
+  age: 23
+});
+
+/**
+ * recursive union types are not supported
+ */
+interface Node {
+  next: Node | null;
+}
+
+declare const nodeCollection: Collection<Node>;
+
+// circular type error is thrown
+// @ts-expect-error
```

Comment:
I talked with Neal about this - `expectError` was unable to catch the recursive type error that's thrown.  `// @ts-expect-error` effectively does the same thing